thriftbp/server.go

```diff
@@ -91,36 +93,145 @@ func NewServer(cfg ServerConfig) (*thrift.TSimpleServer, error) {
 	return server, nil
 }
 
+// NewServerFromOpts creates a new ServerConfig instance and
+// delegates to NewServer so that it can instantiate a
+// new thrift.TSimpleServer.
+func NewServerFromOpts(opts ...ServerOpt) (*thrift.TSimpleServer, error) {
+
+	cfg, err := BuildConfig(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewServer(*cfg)
+}
+
 // NewBaseplateServer returns a new Thrift implementation of a Baseplate
 // server with the given config.
 func NewBaseplateServer(
 	bp baseplate.Baseplate,
 	cfg ServerConfig,
 ) (baseplate.Server, error) {
-	middlewares := BaseplateDefaultProcessorMiddlewares(
-		DefaultProcessorMiddlewaresArgs{
-			EdgeContextImpl:                    bp.EdgeContextImpl(),
-			ErrorSpanSuppressor:                cfg.ErrorSpanSuppressor,
-			ReportPayloadSizeMetricsSampleRate: cfg.ReportPayloadSizeMetricsSampleRate,
-		},
+
+	opts := append(
+		[]ServerOpt{ServerOptFrom(cfg)},
+		DefaultServerOpts(bp)...,
 	)
-	middlewares = append(middlewares, cfg.Middlewares...)
-	cfg.Middlewares = middlewares
-	if cfg.Logger == nil {
+
+	return NewBaseplateServerFromOpts(bp, opts...)
+}
+
+// NewBaseplateServerFromOpts returns a new Thrift implementation of a Baseplate
+// server using a config built from the supplied opts ServerOpt.
+func NewBaseplateServerFromOpts(
+	bp baseplate.Baseplate,
+	opts ...ServerOpt,
+) (baseplate.Server, error) {
+
+	srv, err := NewServerFromOpts(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return ApplyBaseplate(bp, srv), nil
+}
+
+// ServerOpt is a type used for defining configuration arguments
+// needed for creating a baseplate server. It allows returning
+// an error to signal that a configuration value was invalid
+// or failed to be set.
+type ServerOpt func(cfg *ServerConfig) error
+
+// BuildConfig creates a new ServerConfig instance and applies
+// all the supplied configuration options to it.
+func BuildConfig(opts ...ServerOpt) (*ServerConfig, error) {
+	cfg := &ServerConfig{}
+	for _, opt := range opts {
+		err := opt(cfg)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cfg, nil
+}
+
+// ServerOptFrom returns a server option that overwrites the
+// any existing values in a ServerConfig with those defined
+// in src.
+func ServerOptFrom(src ServerConfig) ServerOpt {
```

Comment:
The convention around these names is that they tend to start `With` e.g. `WithConfigFrom(ServerConfig)`.  godoc will sort them under the appropriate type, e.g. DialOption in https://pkg.go.dev/google.golang.org/grpc